[#51][advertisement] feat : 광고 포인트 지급

venus-y opened the pull request:
## Description
- 댓글 도메인으로부터 요청받은 포인트 지급 이벤트를 처리하는 로직 구현
- Kafka, Redis, Redisson 의존성 추가
- 테스트는 곧 추가할 예정

Resolves: #(Isuue Number)

## PR Type
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

venus-y committed:
[#51][advertisement] feat : 광고 포인트 지급

- 댓글 도메인으로부터 요청받은 포인트 지급 이벤트를 처리하는 로직 구현
- Kafka, Redis, Redisson 의존성 추가
- 테스트는 곧 추가할 예정

hp5234 reviewed (approved):
고생하셨습니다.

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/advertisement/application/message/consumer/PointUpdateConsumer.java:
```diff
@@ -0,0 +1,11 @@
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.springframework.kafka.support.Acknowledgment;
+
+public interface PointUpdateConsumer {
```
금성님 해당 부분 리팩토링이 필요할 것 같습니다.
컨슈머 쪽의 경우 우리가 별도의 인터페이스가 필요하지 않으며, 
인프라스트럭처 Consumer쪽에서 응용 계층의 추상화된 Usecase를 통해 응용 계층에 접근 가능하게 할 수 있을 것 같습니다 ㅎㅎ !

```
package com.newpick4u.client.advertisement.application.usecase;

import com.newpick4u.client.advertisement.application.PointUpdateCommand;

public interface PointUpdateUseCase {
    void update(PointUpdateCommand command);
}
```

```
package com.newpick4u.client.advertisement.infrastructure.kafka;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.newpick4u.client.advertisement.application.PointUpdateCommand;
import com.newpick4u.client.advertisement.application.usecase.PointUpdateUseCase;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.kafka.support.Acknowledgment;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
@Slf4j
public class PointUpdateConsumerImpl {

  private final ObjectMapper objectMapper;
  private final PointUpdateUseCase pointUpdateUsecase;

  @KafkaListener(topics = "${spring.kafka.consumer.topics.point-request}",
      groupId = "${spring.kafka.consumer.groups.point-request}",
      containerFactory = "updateMessageConcurrentKafkaListenerContainerFactory"
  )
  public void consume(ConsumerRecord<String, String> record, Acknowledgment acknowledgment)
      throws JsonProcessingException {
    try {
      PointUpdateCommand message = objectMapper.readValue(record.value(), PointUpdateCommand.class);
      pointUpdateUsecase.update(message);
      acknowledgment.acknowledge();
    } catch (Exception e) {
      log.error("Kafka consume error", e);
    }
  }
}

```

```
package com.newpick4u.client.advertisement.application;

import com.newpick4u.client.advertisement.application.dto.request.CreateAdvertiseRequestDto;
import com.newpick4u.client.advertisement.application.usecase.PointUpdateUseCase;

import java.util.UUID;

public interface AdvertisementService extends PointUpdateUseCase { <--- 해당 부분 응용 서비스 AdvertisementService에 PointUpdate에 대한 기능을 추가한다고 보실 수 있겠습니다 ~! (_ _)

  UUID createAdvertisement(CreateAdvertiseRequestDto request);

  void updatePointCounter(PointUpdateCommand message);

}

```

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/advertisement/application/message/consumer/RetryPointUpdateConsumer.java:
```diff
@@ -0,0 +1,9 @@
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.springframework.kafka.support.Acknowledgment;
+
+public interface RetryPointUpdateConsumer {
```
인프라스트럭처에서 PointUpdateUseCase를 통해 응용 계층에 접근할 수 있다면 비슷하게 해당 인터페이스는 필요하지 않으며,
RetryPointUpdateConsumerImpl 내부에서 PointUpdateUseCase 인터페이스를 통해 응용 계층에 접근할 수 있겠습니다 ~ :)

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/advertisement/application/message/producer/PointUpdateProducer.java:
```diff
@@ -0,0 +1,8 @@
+
+import com.newpick4u.client.advertisement.application.message.request.PointUpdateMessage;
+
+public interface PointUpdateProducer {
```
금성님 해당 부분은 응용 계층에서 상호작용이 필요한 이벤트 발행에 대한 부분으로
아래와 같이 구현 후 

```
public interface EventPublisher {
    void publish(PointUpdateCommand message);
}
```
실질적인 인프라스트럭처에서 아래와 같이 구현 가능하겠습니다 ~ :)
```
@Component
@RequiredArgsConstructor
public class PointUpdateProducerImpl implements EventPublisher {

@Override
  public void publish(PointUpdateCommand message) {
    try {
      String parsedMessage = parsedMessage(message);
      kafkaTemplate.send(topic, parsedMessage);
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }
```

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/advertisement/application/usecase/AdvertisementServiceImpl.java:
```diff
@@ -19,22 +22,42 @@
 
+  // ToDo : 고도화 파트에서 파티션 정책 수정으로 인해 변경될 가능성 존재
+  @Transactional
+  @DistributedLock(key = "#message.advertisementId")
```
금성님 AOP 활용한 레디슨 분산처리 잘 고민하신 것 같습니다 👍 

금성님 다만 해당 부분에서 @DistributedLock(key = "#message.advertisementId")
요렇게 전달 후 AOP 쪽 확인해보니 단순히 distributedLock.key() 만 하고 있어서 별도의 SpEL을 통한 값 파싱 부분이 보이지 않는 것 같습니다. ㅎㅎ 해당의 경우 #message.advertisementId 내부의 값을 빼오는게 아니고 "#message.advertisementId" 자체가 스트링 형태의 키가 될 수 있을 것 같습니다 ~!

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/advertisement/infrastructure/kafka/KafkaConfig.java:
```diff
@@ -0,0 +1,108 @@
+
+    DeadLetterPublishingRecoverer recoverer = new DeadLetterPublishingRecoverer(
+        kafkaTemplate,
+        (record, ex) -> new TopicPartition(topic, record.partition()));
```
금성님 해당 부분의 record.partition() 의 경우 DLQ도 기존 토픽의 파티션의 ID를 가져가게 되기 때문에, 
기존 토픽의 파티션이 3개인 경우 DLQ도 동일하게 3개로 구성하셔야 문제가 없을 것 같습니다 :)

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/advertisement/infrastructure/kafka/PointUpdateConsumerImpl.java:
```diff
@@ -0,0 +1,37 @@
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class PointUpdateConsumerImpl implements PointUpdateConsumer {
```
별도로 PointUpdateConsumer를 구현하실 필요가 없을 것 같습니다 ~!!

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/advertisement/infrastructure/kafka/PointUpdateConsumerImpl.java:
```diff
@@ -0,0 +1,37 @@
+public class PointUpdateConsumerImpl implements PointUpdateConsumer {
+
+  private final ObjectMapper objectMapper;
+  private final AdvertisementService advertisementService;
```
해당 부분이 위에서 말씀드린 PointUpdateUseCase 인터페이스를 통해 응용 계층에 접근 가능할 것 같습니다 ~! :)

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/advertisement/infrastructure/kafka/PointUpdateProducerImpl.java:
```diff
@@ -0,0 +1,34 @@
+
+@Component
+@RequiredArgsConstructor
+public class PointUpdateProducerImpl implements PointUpdateProducer {
```
해당 부분은 EventPublisher를 구현하여 publish를 구현해주시면 되겠습니다 ㅎㅎ :) 

현재 구조상으론 produce(..) --> publish(..) 로 바꿔주셔도 될 것 같습니다 :)

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/advertisement/infrastructure/kafka/RetryPointUpdateConsumerImpl.java:
```diff
@@ -0,0 +1,59 @@
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class RetryPointUpdateConsumerImpl implements RetryPointUpdateConsumer {
```
위와 동일합니다. RetryPointUpdateConsumer를 별도로 구현하실 필요가 없을 것 같습니다 :)

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/advertisement/infrastructure/kafka/RetryPointUpdateConsumerImpl.java:
```diff
@@ -0,0 +1,59 @@
+  private static final int INT_BYTE_SIZE = 4;
+  private static final String RETRY_HEADER_KEY = "kafka_deliveryAttempt";
+  private final ObjectMapper objectMapper;
+  private final AdvertisementService advertisementService;
```
해당 부분도 동일하게 UseCase를 통해 접근해주시면 되겠습니다 ~ :)

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/global/aop/DistributedLockAop.java:
```diff
@@ -0,0 +1,59 @@
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class DistributedLockAop {
```
👍 분산 락 AOP구성 너무 잘 선택해주신 것 같습니다 ㅎㅎ

dannyseo9202 commented on client/src/main/java/com/newpick4u/client/global/aop/DistributedLockAop.java:
```diff
@@ -0,0 +1,59 @@
+    MethodSignature signature = (MethodSignature) joinPoint.getSignature();
+    Method method = signature.getMethod();
+
+    String key = REDISSON_LOCK_PREFIX + distributedLock.key();
```
요 부분이 위에서 말씀드린 부분으로 실제 #message.advertisementId 안의 값을 파싱해오는지 확인이 필요할 것 같습니다 ㅎㅎ ~!
SpelExpressionParser와 StandardEvaluationContext 사용해서 가지고 오셔야 할 수도 있습니다 ㅎㅎ

dannyseo9202 reviewed (commented):
리뷰 완료했습니다 금성님 ~!! ㅎㅎ

아래와 같이 계층간 의존성 분리가 완벽한 것 같습니다 ㅎㅎ 👍 👍
![스크린샷 2025-04-11 오전 9 51 24](https://github.com/user-attachments/assets/c409b3e2-02d3-4f59-9ff3-2fd7f2ef26fa)

venus-y committed:
[#51][advertisement] fix : 분산락 적용 관련 수정

- @DistiributeLock 어노테이션을 사용하는 메서드에서 lock의 key로 사용할
값을 해석할 수 있는 CustomSpringELParser를 정의
- 응용계층에 존재하는 updatePointCount 메소드를 포인트 지급 횟수를 갱신한다는 의미를 전달하기 위해 updatePointGrantedCount로 명칭을 변경

venus-y committed:
[#51][advertisement] refactor : dlq 재시도 횟수 변경

- dlq 재시도 횟수 3회 -> 1회로 변경
